### PR TITLE
Group Based Licensing read-only graph changes in beta

### PR DESCRIPTION
**Pull request by SumitParikh**

### CODE REVIEW

File: api-reference/beta/resources/group.md [hunk 1]
```diff
@@ -88,19 +88,23 @@ This resource supports:
...
 |createdDateTime|DateTimeOffset| Timestamp of when the group was created. The value cannot be modified and is automatically populated when the group is created. The Timestamp type represents date and time information using ISO 8601 format and is always in UTC time. For example, midnight UTC on Jan 1, 2014 would look like this: `'2014-01-01T00:00:00Z'`. Read-only. |
 |description|String|An optional description for the group.|
 |displayName|String|The display name for the group. This property is required when a group is created and it cannot be cleared during updates. Supports $filter and $orderby.|
 |groupTypes|String collection| Specifies the type of group to create. Possible values are **Unified** to create an Office 365 group, or **DynamicMembership** for dynamic groups.  For all other group types, like security-enabled groups and email-enabled security groups, do not set this property.|
+|hasMembersWithLicenseErrors|Boolean|true if the group has member or members with license error(s) from this group-based license assignment; otherwise, false. Read-only. Cannot view property. Only supports $filter.|
```
Comment (eketo-msft) [inline api-reference/beta/resources/group.md]: What does this mean?  We don't hide this property from the client, do we?

Comment (SumitParikh) [reply api-reference/beta/resources/group.md]: Cannot view the property by default on group. you have to add a filter to read the value.

**Comment by SumitParikh** [comment]
Cannot view the property by default on group. you have to add a filter to read the value.


Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

File: api-reference/beta/resources/group.md (hunk 1 above)
Comment (eketo-msft) [reply api-reference/beta/resources/group.md]: Do you mean $select?  @lleonard-msft how do we doc these sorts of properties?

Comment (lleonard-msft) [reply api-reference/beta/resources/group.md]: @Lauragra, @angelgolfer-ms: Do you have suggestions here?

File: api-reference/beta/api/group_get.md [hunk 2]
```diff
@@ -48,6 +49,13 @@ GET https://graph.microsoft.com/beta/groups/c28c1cc9-e1ab-4c4d-98d1-d8fdf128b60f
 GET https://graph.microsoft.com/beta/groups/c28c1cc9-e1ab-4c4d-98d1-d8fdf128b60f?$select=description,allowExternalSenders
 ```
 
+To return groups containing members with license errors, use the **$filter** query parameter:
```
Comment (dkershaw10) [inline api-reference/beta/api/group_get.md]: This doesn't seem to be in the right place.  This area is telling me about default properties, and you are throwing a filter example in here.  Suggest that you add this case to the examples section. We have other topics that demonstrate multiple example request/responses.

File: api-reference/beta/resources/group.md [hunk 3]
```diff
@@ -88,6 +88,7 @@ This resource supports:
 | Property	   | Type	|Description|
 |:---------------|:--------|:----------|
 |allowExternalSenders|Boolean|Default is **false**. Indicates if people external to the organization can send messages to the group.|
+|assignedLicenses|[assignedLicense](assignedlicense.md) collection|The licenses that are assigned to the group. Read-only. Not nullable.|
```
Comment (dkershaw10) [inline api-reference/beta/resources/group.md]: You should remove **not-nullable".  It's read-only :)

File: api-reference/beta/resources/licenseAssignmentState.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# licenseAssignmentState resource type
+
+> **Important:** APIs under the /beta version in Microsoft Graph are in preview and are subject to change. Use of these APIs in production applications is not supported.
+
+The **licenseAssignmentStates** property of the [user](user.md) entity is a collection of **licenseAssignmentState**.
```
Comment (dkershaw10) [inline api-reference/beta/resources/licenseAssignmentState.md]: This sentence really tells me nothing.  Essentially you're telling me there's a property that's a collection.  Tell me what it's for and what it represents.

File: api-reference/beta/resources/licenseAssignmentState.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+
+## Properties
+| Property	   | Type	|Description|
+|:---------------|:--------|:----------|
+|assignedByGroup|string|The object id of the group that assigns this license. If the assignment is direct-assigned license, this field will be Null. Read-Only.|
```
Comment (dkershaw10) [inline api-reference/beta/resources/licenseAssignmentState.md]: object id -> id
is direct-assigned license -> is a direct-assigned license

File: api-reference/beta/resources/licenseAssignmentState.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+|disabledPlans|Collection(string)|The service plans that are disabled in this assignment. Read-Only.|
```
Comment (dkershaw10) [inline api-reference/beta/resources/licenseAssignmentState.md]: string -> String

File: api-reference/beta/resources/licenseAssignmentState.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+|error|String|License assignment failure error. If the license is assigned successfully, this field will be Null. Read-Only. Possible values: `CountViolation`, `MutuallyExclusiveViolation`, `DependencyViolation`, `ProhibitedInUsageLocationViolation`, `UniquenessViolation`, and Others.|
```
Comment (dkershaw10) [inline api-reference/beta/resources/licenseAssignmentState.md]: Is there some other place where you explain more about these errors and remediation?  Maybe link to that?

File: concepts/changelog.md [hunk 8]
```diff
@@ -4,6 +4,17 @@ This changelog covers what's changed in Microsoft Graph, including the v1.0 and
 
 For details about known issues with Microsoft Graph APIs, see [Known issues](known_issues.md).
 
+## October 2018
+
+### Directory APIs
+
+| **Change type** | **Version**   | **Description**                          |
+| :-------------- | :------------ | :--------------------------------------- |
+| Addition | Beta | Added the **licenseAssignmentStates** property to the [User](../api-reference/beta/resources/user.md) entity for [Group Based Licensing](https://docs.microsoft.com/azure/active-directory/fundamentals/active-directory-licensing-whatis-azure-portal).|
```
Comment (dkershaw10) [inline concepts/changelog.md]: OK - there is literally NO place that you talk about GBL in any of the topics - it only happens here in the changelog.  **Suggestion**: update the Groups overview topic - https://developer.microsoft.com/en-us/graph/docs/api-reference/beta/resources/groups-overview to have a paragraph about group based licensing and link to your topic on docs.microsoft.com.

Something else you might consider is whether it makes sense to have a Graph conceptual topic on licensing and subscriptions all up?  You might want to talk to @lleonard-msft about this.

**Comment by SumitParikh** [comment]
HI Dan,

This is the guidance I got from Lance. See attached email thread on this. please work with Lance to see if we all are on the same page before I make more changes 😊





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

**Comment by SumitParikh** [comment]
HI Dan,

To my knowledge these error codes are ones we get from office except `ProhibitedInUsageLocationViolation`. @Jimmy Zhu<mailto:Zhenying.Zhu@microsoft.com> please confirm. I don’t believe these are documented somewhere else. @Vineet Fernandes<mailto:vifernan@microsoft.com> – do you have any idea?





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

**Comment by SumitParikh** [comment]
Hi Dan,

Lance and I discussed this and I did replay to thins in attached email. We decided against adding GBL to groups overview as Groups is owned by office and all they talk about is office groups. GBL does not support office groups. Yes your suggestion to create a licensing topic or GBl topic makes sense but Lance told me there is not a single AAD topic at feature level today and writers are working on a long term plan to create that structure. Once that structure is ready we can add licensing/GBL there. For now this is not blocking GBL GA.

Thoughts?




Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

**Comment by SumitParikh** [comment]
Hi,

From my knowledge, those errors means:

CountViolation: The seat count of the subscription is all used up so the user cannot be assigned a license of this subscription.
MutuallyExclusiveViolation: The license contains a service plan that is non-stackable with an existing assigned plan on the user. For example Exchange standard vs. Exchange enterprise.
DependencyViolation: The license contains a service plan that depends on another service plan but the user doesn't have that.
ProhibitedInUsageLocationViolation: The usage location of the user is prohibit by the license.
UniquenessViolation: The user's proxy address after assigning the license is already taken in the tenant.

Here is a public doc for the remediation:
https://docs.microsoft.com/en-us/azure/active-directory/users-groups-roles/licensing-groups-resolve-problems

Thanks,
Jimmy

From: Sumit Parikh <Sumit.Parikh@microsoft.com>
Sent: Thursday, October 18, 2018 1:40 PM
To: microsoftgraph/microsoft-graph-docs <reply@reply.github.com>; microsoftgraph/microsoft-graph-docs <microsoft-graph-docs@noreply.github.com>; Dan Kershaw <dkershaw@microsoft.com>; Jimmy Zhu <Zhenying.Zhu@microsoft.com>; Vineet Fernandes <vifernan@microsoft.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [microsoftgraph/microsoft-graph-docs] Group Based Licensing read-only graph changes in beta (#3167)

HI Dan,

To my knowledge these error codes are ones we get from office except `ProhibitedInUsageLocationViolation`. @Jimmy Zhu<mailto:Zhenying.Zhu@microsoft.com> please confirm. I don't believe these are documented somewhere else. @Vineet Fernandes<mailto:vifernan@microsoft.com> - do you have any idea?





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

**Comment by SumitParikh** [comment]
Thanks Jimmy.

@Dan Kershaw<mailto:dkershaw@microsoft.com> - is it ok if I point to this documentation link which is outside of graph path?





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

From: Jimmy Zhu <Zhenying.Zhu@microsoft.com>
Sent: Thursday, October 18, 2018 3:04 PM
To: Sumit Parikh <Sumit.Parikh@microsoft.com>; microsoftgraph/microsoft-graph-docs <reply@reply.github.com>; microsoftgraph/microsoft-graph-docs <microsoft-graph-docs@noreply.github.com>; Dan Kershaw <dkershaw@microsoft.com>; Vineet Fernandes <vifernan@microsoft.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [microsoftgraph/microsoft-graph-docs] Group Based Licensing read-only graph changes in beta (#3167)

Hi,

From my knowledge, those errors means:

CountViolation: The seat count of the subscription is all used up so the user cannot be assigned a license of this subscription.
MutuallyExclusiveViolation: The license contains a service plan that is non-stackable with an existing assigned plan on the user. For example Exchange standard vs. Exchange enterprise.
DependencyViolation: The license contains a service plan that depends on another service plan but the user doesn't have that.
ProhibitedInUsageLocationViolation: The usage location of the user is prohibit by the license.
UniquenessViolation: The user's proxy address after assigning the license is already taken in the tenant.

Here is a public doc for the remediation:
https://docs.microsoft.com/en-us/azure/active-directory/users-groups-roles/licensing-groups-resolve-problems<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Fazure%2Factive-directory%2Fusers-groups-roles%2Flicensing-groups-resolve-problems&data=04%7C01%7CSumit.Parikh%40microsoft.com%7C5a5456bc37a34c46a10c08d635458f10%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636754970844074914%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-1&sdata=9AcfTdAXUTEVhv%2BPju6ERsWEYSipy556bW7ILAGLUMo%3D&reserved=0>

Thanks,
Jimmy

From: Sumit Parikh <Sumit.Parikh@microsoft.com<mailto:Sumit.Parikh@microsoft.com>>
Sent: Thursday, October 18, 2018 1:40 PM
To: microsoftgraph/microsoft-graph-docs <reply@reply.github.com>; microsoftgraph/microsoft-graph-docs <microsoft-graph-docs@noreply.github.com<mailto:microsoft-graph-docs@noreply.github.com>>; Dan Kershaw <dkershaw@microsoft.com<mailto:dkershaw@microsoft.com>>; Jimmy Zhu <Zhenying.Zhu@microsoft.com<mailto:Zhenying.Zhu@microsoft.com>>; Vineet Fernandes <vifernan@microsoft.com<mailto:vifernan@microsoft.com>>
Cc: Author <author@noreply.github.com<mailto:author@noreply.github.com>>
Subject: RE: [microsoftgraph/microsoft-graph-docs] Group Based Licensing read-only graph changes in beta (#3167)

HI Dan,

To my knowledge these error codes are ones we get from office except `ProhibitedInUsageLocationViolation`. @Jimmy Zhu<mailto:Zhenying.Zhu@microsoft.com> please confirm. I don't believe these are documented somewhere else. @Vineet Fernandes<mailto:vifernan@microsoft.com> - do you have any idea?





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

File: api-reference/beta/resources/groups-overview.md [hunk 9]
```diff
@@ -111,6 +111,10 @@ To learn more about formulating membershipRules, see [Create attribute-based rul
 
 Office 365 groups in Yammer are used to facilitate user collaboration through Yammer posts. This type of group can be returned through a read request, but their posts can't be accessed through the API. When Yammer posts and conversation feeds are enabled on a group, default Office 365 group conversations are disabled. To learn more, see [Yammer developer API docs](https://developer.yammer.com/docs).
 
+## Group-based licensing
+
+Group-based licensing capability can be used to assign one or more product licenses to a Azure AD group. Azure AD ensures that the licenses are assigned to all members of the group. Any new members who join the group are assigned the appropriate licenses. When they leave the group, those licenses are removed. The feature can only be used with security groups, and Office 365 groups that have securityEnabled=TRUE. To learn more about group-based licensing see [here](https://docs.microsoft.com/azure/active-directory/fundamentals/active-directory-licensing-whatis-azure-portal).
```
Comment (dkershaw10) [inline api-reference/beta/resources/groups-overview.md]: a -> an

File: api-reference/beta/resources/group.md [hunk 10]
```diff
@@ -96,11 +97,13 @@ This resource supports:
...
 |mail|String|The SMTP address for the group, for example, "serviceadmins@contoso.onmicrosoft.com". Read-only. Supports $filter.|
 |mailEnabled|Boolean|Specifies whether the group is mail-enabled. If the **securityEnabled** property is also **true**, the group is a mail-enabled security group; otherwise, the group is a Microsoft Exchange distribution group.|
 |mailNickname|String|The mail alias for the group, unique in the organization. This property must be specified when a group is created. Supports $filter.|
 |membershipRule|String|The rule that determines members for this group if the group is a dynamic group (groupTypes contains "**DynamicMembership**"). For more information about the syntax of the membership rule, please refer to [Membership Rules syntax](https://azure.microsoft.com/en-us/documentation/articles/active-directory-accessmanagement-groups-with-advanced-rules/)|
 |membershipRuleProcessingState|String|Indicates whether the dynamic membership processing is on or paused. Possible values are "On" or "Paused"|
+|membersWithLicenseErrors|[User](user.md) collection|A list of group members with license errors from this group-based license assignment. Read-only.|
```
Comment (dkershaw10) [inline api-reference/beta/resources/group.md]: Missed this last time - is this a relationship?
Shouldn't this be under the relationship section?

**Comment by SumitParikh** [comment]
Adding Jimmy to confirm. I believe “membersWithLicenseErrors” is a property and not relationship.





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

**Comment by SumitParikh** [comment]
It is similar to members. Since members is a relationship, I think membersWithLicenseErrors is also a relationship.

Thanks,
Jimmy

From: Sumit Parikh <Sumit.Parikh@microsoft.com>
Sent: Friday, October 26, 2018 8:06 AM
To: microsoftgraph/microsoft-graph-docs <reply@reply.github.com>; microsoftgraph/microsoft-graph-docs <microsoft-graph-docs@noreply.github.com>; Jimmy Zhu <Zhenying.Zhu@microsoft.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [microsoftgraph/microsoft-graph-docs] Group Based Licensing read-only graph changes in beta (#3167)

Adding Jimmy to confirm. I believe "membersWithLicenseErrors" is a property and not relationship.





Thanks and Regards,
Sumit Parikh, Senior Program Manager
Microsoft Corporation
M: +1 425.241.5440

**Pull request merged**